### ISSUE TITLE
Ideal metrics for Monitoring Druid Operator

### ISSUE THREAD

**Issue by cintoSunny** [druid-io/druid-operator#243]
We have multiple metrics emitted from the Druid Operator. Wondering what is the most important set of metrics that we need to monitor to understand the health of the operator, especially since we are watching multiple namespaces. Would love to hear from you guys before adding to the dashboard to verify if we are missing anything.

**Comment by AdheipSingh** [comment]
thanks for opening issue.

You can refer to this https://github.com/druid-io/druid-operator/issues/90#issue-741393427  

IMHO the main metrics too monitor is the worker queue performance. You can try upgrading multiple clusters in parallel and measure the operator performance.

> I also checked the metric success label, and it always comes as 0. I am trying to figure out the reason for this
```ctrlmetrics.ReconcileTotal.WithLabelValues(c.Name, labelSuccess).Inc())```

can you give some more details around this ?

**Comment by cintoSunny** [comment]
Thanks. these are the ones I am thinking of monitoring:
controller_runtime_reconcile_total
controller_runtime_reconcile_errors_total
controller_runtime_reconcile_time_seconds
controller_runtime_reconcile_time_seconds_bucket
controller_runtime_active_workers
workqueue_unfinished_work_seconds

Let me know what you think? Is there is anything I am missing, let me know.

**Comment by cintoSunny** [comment]
> > I also checked the metric success label, and it always comes as 0. I am trying to figure out the reason for this
> > `ctrlmetrics.ReconcileTotal.WithLabelValues(c.Name, labelSuccess).Inc())`
> 
> can you give some more details around this ?

I see that there is a metric 
```controller_runtime_reconcile_total{controller="druid",result="success"}```

I am not sure what this is exactly used for i.e what is the use case for that?

**Comment by AdheipSingh** [comment]
> controller_runtime_reconcile_errors_total
controller_runtime_reconcile_time_seconds
controller_runtime_reconcile_time_seconds_bucket
controller_runtime_active_workers
workqueue_unfinished_work_seconds

LGTM , it would be great if we can publish some benchmark results around the performance.

> Also, is there an easy way to identify the number of druid clusters the operator is watching? I want to add a metric for that as well

This should be helpful. There isn't any metric for it IMO.

**Comment by cintoSunny** [comment]
Sure. Let me know which benchmark results/metrics you have in mind. I can share them. 

> > Also, is there an easy way to identify the number of druid clusters the operator is watching? I want to add a metric for that as well
> 
> This should be helpful. There isn't any metric for it IMO.

I thought so. Do you have any possible solution for this in mind? I can try it out and let you know.

**Comment by AdheipSingh** [comment]
> Sure. Let me know which benchmark results/metrics you have in mind. I can share them.

Nothing concrete as of now, would be just curious to see the workerqueue performance. 

we can add custom metrics to the metrics endpoint /metrics in Prometheus types. 
In case for no of clusters druid handling this can be a gauge. 

it would be nice to have a separate pkg for monitoring with an interface, so we can keep on adding metrics in the future. 
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/metrics - can leverage this
